src/main/surfrdUtilsMod.F90

```diff
@@ -143,6 +145,140 @@ subroutine convert_cft_to_pft( begg, endg, cftsize, wt_cft )
 
   end subroutine convert_cft_to_pft
 
+  !-----------------------------------------------------------------------
+  subroutine collapse_nat_pfts(wt_nat_patch, natpft_size, begg, endg, n_dom_pfts)
+    !
+    ! DESCRIPTION
+    ! Collapse to the top N dominant pfts (n_dom_pfts)
+    ! - Bare ground could be up to 1 patch before collapsing.
+    ! - Pfts could be up to 14 before collapsing if create_crop_landunit = .T.
+    ! - Pfts could be up to 16 before collapsing if create_crop_landunit = .F.
+    !
+    ! !USES:
+    use clm_varpar, only: natpft_lb, natpft_ub, maxveg
+    use pftconMod, only: noveg, nc3crop, nc3irrig, pftcon
+    use array_utils, only: find_k_max_indices
+    !
+    ! !ARGUMENTS:
+    ! Use begg and endg rather than 'bounds', because bounds may not be
+    ! available yet when this is called
+    integer, intent(in) :: begg  ! Beginning grid cell index
+    integer, intent(in) :: endg  ! Ending grid cell index
+    integer, intent(in) :: natpft_size  ! how many pfts 
+    integer, intent(in) :: n_dom_pfts  ! # dominant soil pfts
+    ! These arrays modified in-place
+    ! Weights of pfts per grid cell
+    ! Dimensioned [g, natpft_lb:natpft_ub]
+    real(r8), intent(inout) :: wt_nat_patch(begg:, natpft_lb:)
+    !
+    ! !LOCAL VARIABLES:
+    integer :: g  ! gridcell index
+    integer :: m  ! pft index
+    integer :: n  ! index of the order of the dominant pfts
+    integer, allocatable :: max_indices(:)  ! array of dominant pft index values
+    real(r8) :: wt_dom_nat_sum
+
+    character(len=*), parameter :: subname = 'collapse_nat_pfts'
+
+    !-----------------------------------------------------------------------
+
+    SHR_ASSERT_ALL((ubound(wt_nat_patch) == (/endg, natpft_lb+natpft_size-1/)), errMsg(sourcefile, __LINE__))
+
+    ! Find the top N dominant pfts to collapse the data to
+    ! n_dom_pfts < 0 is not allowed (error check in controlMod.F90)
+    ! Default value n_dom_pfts = 0 or a user-selected n_dom_pfts = natpft_size
+    ! means "do not collapse pfts" and skip over this subroutine's work
+    if (n_dom_pfts > 0 .and. n_dom_pfts < natpft_size) then
+       allocate(max_indices(n_dom_pfts))
+       do g = begg, endg
+          max_indices = 0._r8  ! initialize
```

Comment:
This should be 0 rather than 0._r8. (It should do the same thing in practice, but setting it to 0._r8 makes it look like the variable is a double rather than an int.)

Reply:
But actually, there's a bigger point here: `find_k_max_indices` allocates the `max_indices` array, so the allocation and initialization you do here are essentially undone. However, I see that the allocation of `max_indices` in that utility function was probably a poor design choice, because it leads to this allocation being redone for every grid cell, which is expensive and unnecessary. So I'd recommend:

1. Making the change I suggest above (changing `0._r8` to simply `0`)

2. Changing `find_k_max_indices`:
  - Make `max_indices` not allocatable
  - Add a `shr_assert` that `max_indices` has size `k`
  - Change the documentation: change `The output array (max_indices) is allocated here` to `The output array (max_indices) should be of size k`

3. Changing the unit tests of `find_k_max_indices` (in `test_find_k_max_indices.pf`): in each test routine, rather than having `max_indices` be an allocatable array, make `max_indices` be declared to be the appropriate size. e.g., in `alreadyInOrder`, change `integer, allocatable :: max_indices(:)` to `integer :: max_indices(3)`, and similarly for the other tests (based on the value of `k` in each test).

Let me know if you want me to help with this cleanup, if it's not clear to you.

Reply:
Oops, thank you for catching!

> This should be 0 rather than 0._r8.

Reply:
These changes make sense; however, I have two concerns:
1) `grep 'subroutine find_k_max' */*`
returns one instance of the subroutine in array_utils.F90 and one in array_utils.F90.in. I've noticed the .in files before, but I do not know what they do. Should I be making these changes to both?
2) git status and git diff do not register my mods in array_utils.F90, so I get the sense that these changes are in the externals. I do not know how to commit and push such changes...

> Let me know if you want me to help with this cleanup, if it's not clear to you.

Reply:
@slevisconsulting the *.F90.in files are the source that create the .F90 file of the same name. They are used when you need to do a bunch of code that does the exact same thing, but for different data types (real, integer, etc.). This means you don't have to duplicate so much code.

The reason git status doesn't show anything in array_utils.F90 is because that file is not checked into git, since it's created by the other one.

Reply:
Aha, got it, thank you @ekluzek .